### ISSUE TITLE
Advanced guide / in-depth documentation and explanation

### ISSUE THREAD

**Issue by KingoftheHomeless** [polysemy-research/polysemy#232]
From a recent discussion on [reddit](https://www.reddit.com/r/haskell/comments/db97qw/trouble_choosing_an_effect_library/f1zfeza/), it's clear we need some way to provide a better understanding of the architecture of the library. Here are some points I can think of:

1. What `Tactics`/`Strategy` is, the differences between them, and how to use them. 
    1. Related, what is meant by "effectful state"/"state of the world" in the context of `Tactics`/`Strategy` (although it will be tricky explaining this without straying too far into implementation details).
2. A deeper discussion of local/global state semantics, and how (and potentially, why) `Embed`/`Final`/transformative interpreters don't follow the same rules that pure/`withLowerToIO` interpreters do.
3. What `Final` is, what it provides, and why it is better than having `lower-` interpreters.
    1. Related to 2., better communicate the caveats with `Final`, and how frightened you should be each _Beware_. (Which is, honestly, not at all once you understand how `-Final` interpreters differ with regards to local/global state semantics. I think the _Beware_ attached to `-Final` interpreter is a bit too scary.)
4. The semantics behind more complicated effects and interpreters. `Resource` and `Async` especially; _every_ interpreter for these have traps or subtle behaviour. `asyncToIO` and `resourceToIO` are especially severe, but there's also smaller stuff like the (undocumented) behaviour of `bracketOnError` with `runResource`. 

Most of these points are already partly addressed in the documentation, but need to be covered more in depth, and collectively; many related pieces of info are scattered throughout the docs, and would benefit from being discussed somewhere as a whole (especially local/global state semantics).

The aim is to give a more complete understanding of the library, it's quirks and components, so that people may confidently understand and predict how various interpreters behave and interact. In order not to overwhelm beginners, this should probably amount to some form of guide which is located separately from the documentation itself (but advertised within the documentation).

I can contribute to any writing once we've settled on what we'll do.

**Comment by dhouthoo** [comment]
This! The style of @rkaippully 's tutorial is gold for beginners. If his use case could be extended or  other simple use cases could be found ...

For example, I'm struggling on how to wrap simple database action effects in a transaction effect (commit,savepoint,rollback).

If every of the basic effects and interpreters is documented your half way writing a Polysemy book :-)

**Comment by isovector** [comment]
Thanks for taking charge here, @KingoftheHomeless. 

In my mind, the most important idea to convey is "how to structure things in terms of effects." If you just try to use it as a substitute for `mtl` you're _much better off just using `mtl`._ The deeper philosophy is to make lots of ultra-precise effects, liberally sprinkle them around the codebase, and have applications never explicitly talk about state or errors or any of the effects provided by this library. My fear is that the effects privileged by being in this library will trick people into thinking they should build their app in terms of them, rather than making their own effects.

---

Regarding your points:

I'm pretty sure we can unify `Strategy` and `Tactics`, which would go a lot of way towards resolving some confusion here. Maybe we could put a better name on top too, but I don't have any ideas right now.

(a little offtopic: I think we can get around the deadlocking around `withLowerToIO` --- my original design was that you could use `Tactics` to describe where the computation should run, either on the server thread, or a client. Right now everything runs on the server, which is why it deadlocks. Coming up with a nice API around this would probably get rid of the a lot of the "every interpreter for these have traps or subtle behaviour" stuff)

I agree with the sentiments about "beware." We could instead say "this interpreter provides `Final` semantics", and link to a relevant part of the documentation.

Right now too much documentation is attached to `Sem` itself. Instead I think we should move that to be well-subsectioned parts of documentation on the `Polysemy` module. This would give a good place to put all the documentation, and a convenient place to link to from other haddocks.

**Comment by isovector** [comment]
Just wrote this about some of the philosophy: https://www.reddit.com/r/haskell/comments/db97qw/trouble_choosing_an_effect_library/f23ho0k/

**Comment by jhenahan** [comment]
I think it'd be really helpful to have some sort of guidance on the how and why of writing your own effects, and how to read effect GADTs for new users.

When I see something like `data MyEffect m a`, I take that to be a stateless effect, since there's nowhere for state to live. When I see `data MyEffect s m a`, I think of it as an effect which carries around some state `s`. What if I want to carry around two states? Do I want `data MyEffect s t m a`, or do I just want to interpret the effect twice with a different state param? What else can I do with type parameters other than stick some state in them?

Importantly, I don't know if this is an accurate way to describe these types. I might just be missing some interpretation of those data declarations that's leaving me with an incomplete picture, limiting the kinds of effects I can write for myself.

**Comment by ocharles** [comment]
I don't think it's right to think of `MyEffect m a` as stateless. 

```haskell
data IntState m a where
  Inc :: IntState m ()
  Dec :: IntState m ()
  Current :: IntState m Int
```

clearly has "state", it's just monomorphic.

**Comment by isovector** [comment]
@jhenahan this is a super helpful question, thanks. The kind of the effect GADT _has nothing to do with the statefulness of the effect_ --- statefulness is entirely a property of the interpretation of an effect. For example, I can write an effect:

```haskell
data IntegerMaker m a where
  Make :: IntegerMaker m Int
```

and provide an constant interpreter for it:

```haskell
constMaker v = interpret \case 
  Make -> pure v
```

or I can make an auto incrementing counter for it:

```haskell
autoIncr = evalState 0 . reinterpret \case
  Make -> get <* modify (+1)
```

**Comment by jhenahan** [comment]
Right. I disabused myself of that notion after poking around and writing some less-trivial effects, but early on that was absolutely how I saw them. The above is something I dug out of some notes I wrote while getting started with the library.

So the question is less "are these extra slots state" and more "what's the right way to explain each part of an effect GADT" for someone who's just getting started?

**Comment by TheMatten** [comment]
@jhenahan Personally, I don't see them much differently from  signatures of actual functions - they're no more than descriptions of functions that take some arguments (with types possibly set by some argument to effect's type constructor) and return some result in effects monad. I don't see myself building effects and then interfaces around it - their shape is motivated by some existing APIs and Effect GADTs just encode them so that `polysemy` can understand them.

If there's something to compare them to, it's maybe defunctionalization - they represent use of some function by storing "symbol" and it's arguments, which some "evaluator" can then use to interpret them.

**Comment by sorki** [comment]
Quoting @TheMatten from #342 so it won't get lost:

> difference (between Strategic and Tactical) basically is that Strategic gives you access to final monad (IO) wrapped around opaque state f in threaded > action, which we need to satisfy type of timeouts argument (IO a). On the other hand, Tactical only gives you f a - attempt to lift > it into IO with pure wouldn't help, because action would already be evaluated at that point.
